OrbitSshQt/Session.cpp

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "OrbitSshQt/Session.h"
+
+#include <optional>
+#include <system_error>
+
+#include "OrbitBase/Logging.h"
+#include "OrbitSsh/Error.h"
+#include "OrbitSsh/Session.h"
+#include "OrbitSshQt/Error.h"
+
+namespace OrbitSshQt {
+
+outcome::result<void> Session::startup() {
+  switch (currentState()) {
+    case State::kInitial:
+    case State::kDisconnected: {
+      OUTCOME_TRY(socket, OrbitSsh::Socket::Create());
+      socket_ = std::move(socket);
+      notifiers_.emplace(socket_->GetFileDescriptor());
+      QObject::connect(&notifiers_->read, &QSocketNotifier::activated, this,
+                       &Session::OnEvent);
+      QObject::connect(&notifiers_->write, &QSocketNotifier::activated, this,
+                       &Session::OnEvent);
+      setState(State::kSocketCreated);
+    }
+    case State::kSocketCreated: {
+      OUTCOME_TRY(socket_->Connect(credentials_.host, credentials_.port));
+      setState(State::kSocketConnected);
+    }
+    case State::kSocketConnected: {
+      OUTCOME_TRY(session, OrbitSsh::Session::Create(context_));
+      session_ = std::move(session);
+      session_->SetBlocking(false);
+      setState(State::kSessionCreated);
+    }
+    case State::kSessionCreated: {
+      OUTCOME_TRY(session_->Handshake(&socket_.value()));
+      setState(State::kHandshaked);
+    }
+    case State::kHandshaked: {
+      OUTCOME_TRY(session_->MatchKnownHosts(
+          credentials_.host, credentials_.port, credentials_.known_hosts_path));
+      setState(State::kMatchedKnownHosts);
+    }
+    case State::kMatchedKnownHosts: {
+      OUTCOME_TRY(
+          session_->Authenticate(credentials_.user, credentials_.key_path));
+      setState(State::kConnected);
+    }
+    case State::kStarted:
+    case State::kConnected:
+    case State::kShutdown:
+    case State::kAboutToDisconnect:
+    case State::kDone:
+    case State::kError:
+      break;
+  }
+
+  return outcome::success();
+}
+
+outcome::result<void> Session::shutdown() {
+  switch (currentState()) {
+    case State::kInitial:
+    case State::kDisconnected:
+    case State::kSocketCreated:
+    case State::kSocketConnected:
+    case State::kSessionCreated:
+    case State::kHandshaked:
+    case State::kMatchedKnownHosts:
+    case State::kStarted:
+    case State::kConnected:
+      break;
```

Comment:
these are not supposed to happen right? maybe log a error or fail?

Reply:
I added FATAL calls in all these switch statements.